core/src/transport.rs

```diff
@@ -58,6 +58,30 @@ pub use self::upgrade::Upgrade;
 
 static NEXT_LISTENER_ID: AtomicUsize = AtomicUsize::new(1);
 
+/// The port use policy for a new connection.
+#[derive(Debug, Copy, Clone, Default, PartialEq, Eq, Hash)]
+pub enum PortUse {
+    /// Always allocate a new port for the dial.
+    New,
+    /// Best effor reusing of an existing port.
+    ///
+    /// If there is no listener present that can be used to dial, a new port is allocated.
+    #[default]
+    Reuse,
+}
+
+/// Options to customize the behaviour during dialing. Through it one can specify:
+/// - The role of the endpoint (dialer or listener).
+/// - The port use policy.
+#[derive(Debug, Copy, Clone)]
+pub struct DialOpts {
+    /// The endpoint establishing a new connection. This option removes the necessity for the
+    /// `dial_as_listener` API.
```

Comment:
No need to talk about old APIs.

```suggestion
    /// The endpoint establishing a new connection.
    ///
    /// When attempting a hole-punch, both parties simultaneously "dial" each other but one party has to be the "listener" on the final connection.
    /// This option specifies the role of this node in the final connection.
```